File: Math/intToRoman.cpp
/*
* #: 12
* Title: Integer to Roman
* Description:
* ------
* Given an integer, convert it to a roman numeral.
*
*
* Input is guaranteed to be within the range from 1 to 3999.
* ------
* Time: O(n)
* Space: O(1)
* Difficulty: Medium
*/
#include <iostream>
#include <vector>
#include <string>
#include <cassert>
using namespace std;


class Solution {
public:
    string intToRoman(int num) {
        vector<string> th {"", "M", "MM", "MMM"};
        vector<string> hu {"", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"};
        vector<string> te {"", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"};
        vector<string> di {"", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"};
        string out = th[num/1000] + hu[(num%1000)/100] + te[(num%100)/10] + di[num%10];
        return out;
    }
};


int main(void)
{
    Solution sol = Solution();
    int num = 3989;
    cout << sol.intToRoman(num) << endl;
    assert(sol.intToRoman(num) == "MMMCMXCIX");
    return 0;
}

File: Math/intToRoman.py
"""
#: 12
Title: Integer to Roman
Description:
------
Given an integer, convert it to a roman numeral.


Input is guaranteed to be within the range from 1 to 3999.
------
Time: O(n)
Space: O(1)
Difficulty: Medium
"""
class Solution(object):
    def intToRoman(self, num):
        """
        :type num: int
        :rtype: str
        """
        th = ["", "M", "MM", "MMM"]
        hu = ["", "C", "CC", "CCC", "CD", "D", "DC", "DCC", "DCCC", "CM"]
        te = ["", "X", "XX", "XXX", "XL", "L", "LX", "LXX", "LXXX", "XC"]
        di = ["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX"]
        res = th[num/1000] + hu[(num%1000)/100] + te[(num%100)/10] + di[num%10]
        return res


if __name__ == '__main__':
    sol = Solution()
    num = 3999
    assert sol.intToRoman(num) == "MMMCMXCIX", "Not correct!"

